### PR TITLE
Add Workload Identity Federation

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context
  _This is part 1 of the migration to WIF. Part 2 requires us to change the Google ruby library for making requests to BigQuery and that will follow this PR. The documentation references some of the changes that will appear in that PR._ 

      Add Workload Federated Identity

      Add classes and configuration to enable pods to request Service
      Account Impersonation Tokens to make authenticated requests to Google
      Cloud.

      These classes will be used to replace the use of Google JSON
      credentials in order to harden security around our access to BigQuery.

      The change consists of A new namespace WorkloadIdentityFederation which
      includes the configurable module.

      The module is configured in the `workload_identity_federation.rb` initializer.

      The contingent classes require instantiation though the
      UserCredentials class uses a class instance variable to store the
      latest retrieved Google Credentials in memory.

      The implementation uses Faraday as it is considered our default HTTP
      client within the application.

      Included is a testing module `WorkloadIdentityFederationStubs` to stub the requests
      of each contingent class as well as stub the Azure configuration.

## Changes proposed in this pull request

Add library classes to allow for the application to request Google Credentials via Workload Identity Federation.

## Guidance to review

Add library classes to allow for the application to request Google Credentials via Workload Identity Federation.

#### How to verify

This feature will only work on the `worker` or `secondary-worker` pods so you'll need to adjust the Makefile to access those pods.

I tend to edit the Makefile to change the pod that the `make env shell` command accesses.

Change the line that matches this like so:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/Makefile#L182

```
-	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-clock-worker), $(eval APP_NAME=apply-${APP_ENV}-clock-worker))
+	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-worker), $(eval APP_NAME=apply-${APP_ENV}-worker))
```

Then use `PR_NUMBER=9689 make review_aks shell` 

You can also inspect the environment variables on the pod with `PR_NUMBER=9689 make review_aks view-app-secrets` to see that `GOOGLE_CLOUD_CREDENTIALS_STATS` is set.

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)
- [ ] Inform data insights team due to database changes
- [x] Make sure all information from the Trello card is in here
- [x] Rebased main
- [ ] Cleaned commit history
- [ ] Tested by running locally
- [x] Add PR link to Trello card

### CODE REVIEW

**Comment by inulty-dfe** [comment]
[BigQuery: Move Monthly Statistics processes to Workload Identity Federation (WIF) authentication](https://trello.com/c/NYxhoYWi/2024-bigquery-move-monthly-statistics-processes-to-workload-identity-federation-wif-authentication)

File: lib/azure/access_token.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      @conn = Faraday.new do |b|
+        b.response :json
+        b.request :url_encoded
+      end
+    end
+
+    def call
+      azure_token_response = @conn.get(@url) do |req|
+        req.body = request_body
+      end
```
Comment (inulty-dfe) [inline lib/azure/access_token.rb:14]: This is a weird one. We're making a get request but we're sending a request body. This is how I was able to do it with Faraday.

By default, the second arg to GET requests is query params, and the second arg to POST request is request body. I have to yield a block param to set the request body on a GET request.

**Review by CatalinVoineag — APPROVED** [review]
Looks good, nice 👍🏻

**Review by elceebee — COMMENTED** [review]
I'm struggling to know how to test this locally or in the review app since we can't call these classes from the rails console (the env vars are stored on the workers only). Could you please give some instructions -- maybe kubernetes incantations that I don't have committed to memory?

**Comment by inulty-dfe** [comment]
> I'm struggling to know how to test this locally or in the review app since we can't call these classes from the rails console (the env vars are stored on the workers only). Could you please give some instructions -- maybe kubernetes incantations that I don't have committed to memory?

@elceebee 

I tend to edit the Makefile to change the pod that the `make env shell` command accesses.

Change the line that matches this like so:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/Makefile#L182

```
-	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-clock-worker), $(eval APP_NAME=apply-${APP_ENV}-clock-worker))
+	$(if ${APP_NAME_SUFFIX}, $(eval APP_NAME=apply-${APP_NAME_SUFFIX}-worker), $(eval APP_NAME=apply-${APP_ENV}-worker))
```


Then use `PR_NUMBER=9689 make review_aks shell` 

You can also inspect the environment variables on the pod with `PR_NUMBER=9689 make review_aks view-app-secrets` to see that `GOOGLE_CLOUD_CREDENTIALS_STATS` is set.

**Review by elceebee — APPROVED** [review]

File: config/initializers/azure.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+Rails.application.config.to_prepare do
+  Azure.configure do |config|
+    begin
+      google_cloud_credentials = JSON.parse(ENV.fetch('GOOGLE_CLOUD_CREDENTIALS_STATS', '{}')).deep_symbolize_keys
+    rescue TypeError
+      raise Azure::GoogleCloudCredentialsError
```
Comment (tomas-stefano) [inline config/initializers/azure.rb:6]: Can we add a message here for any future us? "No ENV['GOOGLE_CLOUD_CREDENTIALS_STATS'] found."

**Review by tomas-stefano — APPROVED** [review]
Minor comment, but everything looks good.

Tested on the review app:

```
Publications::ITTMonthlyReportGenerator.new.to_h
```
And it worked.

**Comment by inulty-dfe** [comment]
> Minor comment, but everything looks good.
> 
> Tested on the review app:
> 
> ```
> Publications::ITTMonthlyReportGenerator.new.to_h
> ```
> 
> And it worked.
@tomas-stefano 
This is not testing the code that was added. This is just to Add `Azure::UserCredentials.call` which returns an authentication object for future use. The Statistics are still using the old JSON credentials.

**Comment by tomas-stefano** [comment]
> > Minor comment, but everything looks good.
> > Tested on the review app:
> > ```
> > Publications::ITTMonthlyReportGenerator.new.to_h
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And it worked.
> 
> @tomas-stefano This is not testing the code that was added. This is just to Add `Azure::UserCredentials.call` which returns an authentication object for future use. The Statistics are still using the old JSON credentials.

Humm I assumed it was somehow. Apologies. So using those in the places we use bigquery will be done in a follow up PR?

**Comment by inulty-dfe** [comment]
> > > Minor comment, but everything looks good.
> > > Tested on the review app:
> > > ```
> > > Publications::ITTMonthlyReportGenerator.new.to_h
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And it worked.
> > 
> > 
> > @tomas-stefano This is not testing the code that was added. This is just to Add `Azure::UserCredentials.call` which returns an authentication object for future use. The Statistics are still using the old JSON credentials.
> 
> Humm I assumed it was somehow. Apologies. So in the places we use bigquery will be done in a follow up PR?

Yes, it's outlined in the PR description above

File: config/initializers/azure.rb (hunk 2 above)
Comment (inulty-dfe) [reply config/initializers/azure.rb:6]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/9689/files#diff-a8f8614a7d8204d996de735cb5a9cd44b12f9477f8df6e3411b16df187a6e450R17-R21

**Comment by inulty-dfe** [comment]
My only issue with this PR is that maybe the `Azure` module should be named `Wif`.

**Comment by tomas-stefano** [comment]
> My only issue with this PR is that maybe the `Azure` module should be named `Wif`.

I'd prefer to be called `WorkloadFederatedIdentity` but this is my personal preference.

**Review by tomas-stefano — APPROVED** [review]

**Pull request merged**